src/components/Autocomplete.tsx

```diff
@@ -0,0 +1,141 @@
+/* eslint-disable react-hooks/exhaustive-deps */
+import {
+  useCallback,
+  useMemo,
+  useState,
+  useEffect,
+  useRef,
+} from 'react';
+import classNames from 'classnames';
+import debounce from 'lodash.debounce';
+import { Person } from '../types/Person';
+
+interface Props {
+  people: Person[];
+  delay: number;
+  onSelected: (person: Person | null) => void;
+}
+
+enum Key {
+  Escape = 'Escape',
+}
+
+export const Autocomplete: React.FC<Props> = ({
+  people,
+  delay,
+  onSelected,
+}) => {
+  const [query, setQuery] = useState('');
+  const [appliedQuery, setAppliedQuery] = useState('');
+  const [showDrop, setShowDrop] = useState(false);
+  const [isUserTyping, setIsUserTyping] = useState(false);
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const applyQuery = useCallback(
+    debounce((newQuery: string) => {
+      setAppliedQuery(newQuery);
+      setIsUserTyping(false);
+    }, delay),
+    [setAppliedQuery, delay],
+  );
+
+  const filteredPeople = useMemo(() => {
+    const preparedAppliedQuery = appliedQuery.toLowerCase().trim();
+
+    return people.filter(person => (
+      person.name.toLowerCase().includes(preparedAppliedQuery)
+    ));
+  }, [appliedQuery, people]);
+
+  const handleItemChange = (
+    event: React.MouseEvent<HTMLAnchorElement, MouseEvent>,
+    person: Person,
+  ) => {
+    event.preventDefault();
+    onSelected(person);
+    setShowDrop(false);
+    setQuery(person.name);
+
+    if (inputRef.current) {
+      inputRef.current.blur();
+    }
+  };
+
+  const handleQueryChange = (
+    event: React.ChangeEvent<HTMLInputElement>,
+  ) => {
+    setQuery(event.target.value);
+    applyQuery(event.target.value);
+    setIsUserTyping(true);
+
+    if (event.target.value === '') {
+      onSelected(null);
+    }
+  };
+
+  const handleEscape = useCallback(
+    (event: KeyboardEvent) => {
+      if (event.key === Key.Escape) {
+        setShowDrop(false);
+      }
+    }, [],
+  );
+
+  useEffect(() => {
+    document.addEventListener('keydown', handleEscape);
+
+    return () => {
+      document.removeEventListener('keydown', handleEscape);
+    };
+  }, [handleEscape]);
+
+  return (
+    <div
+      className={classNames(
+        'dropdown',
+        { 'is-active': showDrop && !isUserTyping },
+      )}
+    >
+      <div className="dropdown-trigger">
+        <input
+          ref={inputRef}
+          type="text"
+          placeholder="Enter a part of the name"
+          className="input"
+          value={query}
+          onChange={handleQueryChange}
+          onFocus={() => setShowDrop(true)}
```

Comment:
It's a bad practice to create jsx anonymous function, it's better to create separate function for this
```suggestion
          onFocus={handleOnFocus}
```
Check other places